File: src/protobuf/stream_test.py
# Copyright (c) 2014, Salesforce.com, Inc.  All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
# - Redistributions of source code must retain the above copyright
#   notice, this list of conditions and the following disclaimer.
# - Redistributions in binary form must reproduce the above copyright
#   notice, this list of conditions and the following disclaimer in the
#   documentation and/or other materials provided with the distribution.
# - Neither the name of Salesforce.com nor the names of its contributors
#   may be used to endorse or promote products derived from this
#   software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE
# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
# OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
# TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
# USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

from nose.tools import assert_equal
from pomagma.protobuf.stream import protobuf_stream_dump
from pomagma.protobuf.stream import protobuf_stream_load
from pomagma.util import in_temp_dir


def test_protobuf_stream():
    for filetype in ['', '.gz', '.bz2']:
        yield _test_protobuf_stream, filetype


def _test_protobuf_stream(filetype):
    filename = 'test.stream' + filetype
    expected = ['asdf', '', 'asdfasdfasdf', 'a', 's', '', '', '', 'd', 'f']
    with in_temp_dir():
        print 'dumping'
        protobuf_stream_dump(expected, filename)
        print 'loading'
        actual = list(protobuf_stream_load(filename))
    assert_equal(actual, expected)

File: src/protobuf/stream_test.cpp
#include "stream.hpp"
#include <cstdio>

using namespace pomagma;

void test_dump_load (const char * filename)
{
    std::vector<std::vector<char>> expected = {
        {},
        {'a'},
        {'b', 'c'},
        {},
        {'d'}
    };
    POMAGMA_INFO("dumping to " << filename);
    protobuf_stream_dump(expected, filename);
    POMAGMA_INFO("loading from " << filename);
    auto actual = protobuf_stream_load<std::vector<char>>(filename);
    std::remove(filename);
    POMAGMA_ASSERT_EQ(actual, expected);
}

int main ()
{
    Log::Context log_context("Running Protobuf Stream Test");

    test_dump_load("/tmp/protobuf_stream_test.pbs");
    test_dump_load("/tmp/protobuf_stream_test.pbs.gz");

    return 0;
}
